src/main/java/org/springframework/data/couchbase/repository/query/CouchbaseQueryExecution.java

```diff
@@ -0,0 +1,305 @@
+	 *
+	 * @author Oliver Gierke
+	 * @author Christoph Strobl
+	 * @since 1.5
+	 */
+	final class SlicedExecution implements CouchbaseQueryExecution {
+
+		private final ExecutableFindByQueryOperation.ExecutableFindByQuery<?> find;
+		private final Pageable pageable;
+
+		public SlicedExecution(ExecutableFindByQueryOperation.ExecutableFindByQuery find, Pageable pageable) {
+
+			Assert.notNull(find, "Find must not be null!");
+			Assert.notNull(pageable, "Pageable must not be null!");
+
+			this.find = find;
+			this.pageable = pageable;
+		}
+
+		/*
+		 * (non-Javadoc)
+		 * @see org.springframework.data.couchbase.repository.query.CouchbaseQueryExecution#execute(org.springframework.data.couchbase.core.query.Query)
+		 */
+		@Override
+		@SuppressWarnings({ "unchecked", "rawtypes" })
+		public Object execute(Query query, Class<?> type, String collection) {
+
+			int pageSize = pageable.getPageSize();
+
+			// Apply Pageable but tweak limit to peek into next page
+			Query modifiedQuery = query.with(pageable).limit(pageSize + 1);
+			List result = find.matching(modifiedQuery).all();
+
+			boolean hasNext = result.size() > pageSize;
+
+			return new SliceImpl<Object>(hasNext ? result.subList(0, pageSize) : result, pageable, hasNext);
+		}
+	}
+
+	/**
+	 * {@link CouchbaseQueryExecution} for pagination queries.
+	 *
+	 * @author Oliver Gierke
+	 * @author Mark Paluch
+	 * @author Christoph Strobl
+	 */
+	final class PagedExecution<FindWithQuery> implements CouchbaseQueryExecution {
+
+		private final ExecutableFindByQueryOperation.ExecutableFindByQuery<?> operation;
+		private final Pageable pageable;
+
+		public PagedExecution(ExecutableFindByQueryOperation.ExecutableFindByQuery<?> operation, Pageable pageable) {
+
+			Assert.notNull(operation, "Operation must not be null!");
+			Assert.notNull(pageable, "Pageable must not be null!");
+
+			this.operation = operation;
+			this.pageable = pageable;
+		}
+
+		/*
+		 * (non-Javadoc)
+		 * @see org.springframework.data.couchbase.repository.query.CouchbaseQueryExecution#execute(org.springframework.data.couchbase.core.query.Query)
+		 */
+		@Override
+		public Object execute(Query query, Class<?> type, String collection) {
+
+			int overallLimit = query.getLimit();
+
+			ExecutableFindByQueryOperation.TerminatingFindByQuery<?> matching = operation.matching(query);
+
+			// Apply raw pagination
+			query.with(pageable);
+
+			// Adjust limit if page would exceed the overall limit
+			if (overallLimit != 0 && pageable.getOffset() + pageable.getPageSize() > overallLimit) {
+				query.limit((int) (overallLimit - pageable.getOffset()));
+			}
+
+			return PageableExecutionUtils.getPage(matching.all(), pageable, () -> {
+
+				long count = operation.matching(query.skip(-1).limit(-1)).count();
+				return overallLimit != 0 ? Math.min(count, overallLimit) : count;
+			});
+		}
+	}
+
+	/*
+		final class ReactivePagedExecution<FindWithQuery> implements CouchbaseQueryExecution { // TODO ??
```

Comment:
There's pretty much commented-out code blocks in here.

Reactive pagination returns typically a `Flux<T>`. An alternative approach can be  `Mono<Page<T>>.`